File: backend/app/routers/issue_stats.py
"""Issue stats router: per-org and per-repo issue metrics.

Provides endpoints for viewing issue opened/closed counts and average
time-to-close grouped by organization and repository.
"""

from __future__ import annotations

from typing import Any

import structlog
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.deps import AuthenticatedUser, get_db, require_role
from app.services import issue_stats_service

logger = structlog.get_logger(__name__)

router = APIRouter(prefix="/issue-stats", tags=["issue-stats"])


@router.get("/by-org")
async def get_issue_stats_by_org(
    window_days: int = Query(default=30, ge=1, le=365),
    org: str | None = Query(None, description="Filter to a specific organization"),
    current_user: AuthenticatedUser = Depends(require_role(["analyst", "sys_admin"])),
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    """Return issue opened/closed counts and avg time-to-close grouped by org."""
    rows = await issue_stats_service.get_issue_stats_by_org(
        db,
        window_days=window_days,
        org=org,
    )

    total_opened = sum(r["opened"] for r in rows)
    total_closed = sum(r["closed"] for r in rows)

    return {
        "window_days": window_days,
        "total_opened": total_opened,
        "total_closed": total_closed,
        "orgs": rows,
    }


@router.get("/by-repo")
async def get_issue_stats_by_repo(
    window_days: int = Query(default=30, ge=1, le=365),
    org: str | None = Query(None, description="Filter to a specific organization"),
    current_user: AuthenticatedUser = Depends(require_role(["analyst", "sys_admin"])),
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    """Return issue opened/closed counts and avg time-to-close grouped by org and repo."""
    rows = await issue_stats_service.get_issue_stats_by_repo(
        db,
        window_days=window_days,
        org=org,
    )

    total_opened = sum(r["opened"] for r in rows)
    total_closed = sum(r["closed"] for r in rows)

    return {
        "window_days": window_days,
        "total_opened": total_opened,
        "total_closed": total_closed,
        "repos": rows,
    }

File: frontend/src/api/issueStats.ts
import { api } from './client';

export interface OrgIssueStat {
  org: string;
  opened: number;
  closed: number;
  net_open: number;
  avg_hours_to_close: number | null;
}

export interface RepoIssueStat {
  org: string;
  repo: string;
  opened: number;
  closed: number;
  net_open: number;
  avg_hours_to_close: number | null;
}

export interface IssueStatsByOrgResponse {
  window_days: number;
  total_opened: number;
  total_closed: number;
  orgs: OrgIssueStat[];
}

export interface IssueStatsByRepoResponse {
  window_days: number;
  total_opened: number;
  total_closed: number;
  repos: RepoIssueStat[];
}

export function getIssueStatsByOrg(params: { window_days?: number; org?: string }) {
  return api.get<IssueStatsByOrgResponse>('/issue-stats/by-org', params);
}

export function getIssueStatsByRepo(params: { window_days?: number; org?: string }) {
  return api.get<IssueStatsByRepoResponse>('/issue-stats/by-repo', params);
}
